### PR TITLE
Add DNS provider for reg.ru

### LINKED ISSUE #1074: Add support for reg.ru
Will it be possible to add?
API - https://www.reg.ru/support/help/api2

### PR DESCRIPTION
**Pull request by ldez**
Add DNS provider for reg.ru

API:
- https://www.reg.ru/support/help/api2
- https://www.reg.ru/support/help/api2#zone_remove_record
- https://www.reg.ru/support/help/api2#zone_add_txt

Fixes #1074

### CODE REVIEW

**Comment by ldez** [comment]
https://github.com/go-acme/lego/issues/1074#issuecomment-596699638

```console
2020/03/09 21:12:15 [INFO] [*._my_domain_, _my_domain_] acme: Obtaining bundled SAN certificate
2020/03/09 21:12:16 [INFO] [*._my_domain_] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/...
2020/03/09 21:12:16 [INFO] [_my_domain_] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/...
2020/03/09 21:12:16 [INFO] [_my_domain_] acme: authorization already valid; skipping challenge
2020/03/09 21:12:16 [INFO] [*._my_domain_] acme: use dns-01 solver
2020/03/09 21:12:16 [INFO] [*._my_domain_] acme: Preparing to solve DNS-01
2020/03/09 21:12:16 [INFO] [*._my_domain_] acme: Trying to solve DNS-01
2020/03/09 21:12:16 [INFO] [*._my_domain_] acme: Checking DNS record propagation using [127.0.0.1:53]
2020/03/09 21:12:16 [INFO] Wait for propagation [timeout: 2h0m0s, interval: 10s]
2020/03/09 21:12:16 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:12:26 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:12:36 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:12:46 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:12:56 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:13:06 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:13:16 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:13:26 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:13:36 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:13:47 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:13:57 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:14:07 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:14:17 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:14:27 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:14:37 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:14:47 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:14:57 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:15:07 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:15:17 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:15:27 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:15:37 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:15:47 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:15:57 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:16:07 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:16:17 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:16:27 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:16:37 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:16:47 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:16:57 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:17:07 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:17:17 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:17:27 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:17:37 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:17:47 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:17:57 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:18:07 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:18:17 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:18:27 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:18:37 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:18:48 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:18:58 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:19:08 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:19:18 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:19:28 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:19:38 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:19:48 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:19:58 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:20:08 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:20:18 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:20:28 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:20:38 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:20:48 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:20:58 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:21:08 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:21:18 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:21:28 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:21:38 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:21:48 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:21:58 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:22:08 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:22:18 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:22:28 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:22:38 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:22:48 [INFO] [*._my_domain_] acme: Waiting for DNS record propagation.
2020/03/09 21:23:05 [INFO] [*._my_domain_] The server validated our request
2020/03/09 21:23:05 [INFO] [*._my_domain_] acme: Cleaning DNS-01 challenge
2020/03/09 21:23:05 [INFO] [*._my_domain_, _my_domain_] acme: Validations succeeded; requesting certificates
2020/03/09 21:23:06 [INFO] [*._my_domain_] Server responded with a certificate.

```

File: providers/dns/regru/internal/model.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+const successResult = "success"
+
+// APIResponse is the representation of an API response.
+type APIResponse struct {
+	Result string `json:"result"`
+
+	Answer *Answer `json:"answer,omitempty"`
+
+	ErrorCode   string            `json:"error_code,omitempty"`
+	ErrorParams map[string]string `json:"error_params,omitempty"`
```
Comment (dmke) [inline providers/dns/regru/internal/model.go]: `ErrorParams` seems to be unused.

File: providers/dns/regru/internal/model.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+// DomainResponse is the representation of an API response answer domain.
+type DomainResponse struct {
+	Result string `json:"result"`
+
+	DName string `json:"dname"`
+
+	ErrorCode   string            `json:"error_code,omitempty"`
+	ErrorParams map[string]string `json:"error_params,omitempty"`
```
Comment (dmke) [inline providers/dns/regru/internal/model.go]: Same here.

File: providers/dns/regru/internal/model.go (hunk 1 above)
Comment (ldez) [reply providers/dns/regru/internal/model.go]: The `ErrorsParams` contaings not really useful information:

```json
{
   "charset" : "utf-8",
   "error_code" : "NO_AUTH",
   "error_params" : {
      "command_name" : "zone/remove_record"
   },
   "error_text" : "No authorization mechanism selected",
   "messagestore" : null,
   "result" : "error"
}
```

So I think we will ignore it for now.

Comment (dmke) [reply providers/dns/regru/internal/model.go]: Yeah, fine with me.

**Pull request merged**